[#46] Use similar to identify character-level diffs, then bold those changes

nk9 opened the pull request:
Tests still pass, but not sure if you'd like some other ones for this.

Note that this does add a dependency on [similar](https://crates.io/crates/similar).

nk9 committed:
[#46] Use similar to identify character-level diffs, then bold those changes

nk9 commented:
I want to request a code review from @swolchok. I tried to follow [these instructions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review), but I don't see any Reviewers in the sidebar, so I'm not able to. Hope this will suffice!

swolchok commented on src/terminal.rs:
```diff
@@ -14,43 +14,79 @@
         .map(|(w, h)| (w as usize, h as usize))
 }
+
+pub fn print_bolded_diff_to_terminal(before: &str, after: &str) -> Result<()> {
```
the terminal module is meant to be only about fiddling with the terminal portably, as in the original codemod tool. I would rather see this function in main.rs, with appropriate interface added to (or not deleted from, in the case of `fg`) terminal.rs. it's been a long time since I thought about this in detail, though; feel free to explain if that would be particularly onerous.

swolchok commented on src/main.rs:
```diff
@@ -524,19 +523,13 @@ impl Fastmod {
     }
 
     fn print_diff<'a>(&mut self, diffs: &[DiffResult<&'a str>]) {
```
`DiffResult` comes from the `diff` crate. Does it really make sense for us to be using both `diff` and `similar` for diffing? Perhaps cleaning this up further is out of scope for this PR, but it seems like we should at least open an issue to get back to having only one diffing crate?

swolchok commented:
also, thanks, I can see the bolding in the tests and this looks nice! (this change causes output to be echoed to the terminal during cargo test, but we already weren't being very picky about that: we were emitting a bunch of blank space anyway.)

nk9 commented:
Thanks for the quick reply! I'll have a look tomorrow and see about addressing your requests.

nk9 committed:
Fix excess lines in `cargo test` output.

nk9 committed:
Ignore .idea files.

nk9 committed:
Reserve terminal.rs for crossterm-related code.

nk9 committed:
Switch entirely from `diff` to `similar`.

nk9 commented:
OK, I think I've addressed everything. The `terminal` module is now just a wrapper for `crossterm`. Note that I've switched from `execute!` to `print!` in most cases because this plays nice with the automatic "hide stdout during tests unless there's an error" behavior, so `cargo test` now looks as it should.

I've also entirely removed use of `diff`. `similar` isn't quite as ergonomic as `diff` at the line level, but it's not too bad. Note that diff results now include a newline, so I've had to adjust the test cases around that. It may be possible to change this by adjusting [`newline_terminated`](https://docs.rs/similar/2.7.0/similar/struct.TextDiffConfig.html#method.newline_terminated), but I haven't played around with that.

I have only tested this with UTF-8 files containing newlines. Probably there should be some more tests with more challenging input, but LMK if you think that's required to merge this PR.

Oh, also: I had AI convert the `diffs_to_print` method to use `Change`. It seems to have mostly kept the logic as-is, and I made a few tweaks myself… but I'm not as confident in my understanding of the second half of the function, so please do take a look at that!

nk9 committed:
Refactor to DRY print_bolded_diff

nk9 committed:
Reduce unnecessary diffs due to line reordering.

swolchok commented on src/terminal.rs:
```diff
@@ -14,13 +14,12 @@
+use crossterm::style::ResetColor;
+use crossterm::style::Attribute;
 
+#[derive(Clone)]
```
this should be `#[derive(Clone, Copy)]` and then you don't need a clone() at the `fg` callsite

swolchok commented on src/main.rs:
```diff
@@ -475,72 +473,104 @@ impl Fastmod {
     }
 
+
+    fn print_bolded_diff(&self, before: &str, after: &str) {
```
I nearly made some efficiency complaints about this code (why two passes, why are we hammering on setting the color to bold) because I had missed that this is printing specifically the character level diff of two changed lines, *not* a general diff between two blobs of text. I think that you could make that clearer just by renaming the function to `print_bolded_line_diff` and/or suffixing `before` and `after` with `_line`. I'll leave it up to you what makes the most sense.

nk9 replied:
Good point, I think I've clarified things now.

swolchok commented on src/terminal.rs:
```diff
@@ -36,21 +35,42 @@ impl Color {
 
 pub fn fg(color: Color) {
-    let _ = io::stdout().execute(SetForegroundColor(color.to_crossterm_color()));
+        print!("{}", SetForegroundColor(color.to_crossterm_color()));
```
please `cargo fmt`

swolchok reviewed (approved):
looks good! just a couple minor things please

nk9 committed:
Derive Copy & Clone

nk9 committed:
Clarify wording when printing bolded diff lines.

nk9 committed:
Run cargo fmt

nk9 commented:
I've discovered two issues with this code. First, when the change happens on the final line and the file doesn't end with a newline, you get this:

<img width="1485" height="145" alt="image" src="https://github.com/user-attachments/assets/97e9cd01-c264-40ce-b4ae-91c4094b06d4" />

The other issue is that, unless the current change is on the last line, the last line of the file is not shown in the preview. I have a fix for both issues ready to go. Should I push another commit to this branch, or do you want a new PR?

swolchok commented:
> I've discovered two issues with this code. First, when the change happens on the final line and the file doesn't end with a newline, you get this:
> 
> <img alt="image" width="1485" height="145" src="https://private-user-images.githubusercontent.com/3646730/500533138-97e9cd01-c264-40ce-b4ae-91c4094b06d4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAzNzk4NDAsIm5iZiI6MTc2MDM3OTU0MCwicGF0aCI6Ii8zNjQ2NzMwLzUwMDUzMzEzOC05N2U5Y2QwMS1jMjY0LTQwY2UtYjRhZS05MWM0MDk0YjA2ZDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMTNUMTgxOTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWQzNTc0MzI0YjkzYzQwODhkMzA3MDUyNDdkZDE0ZjZlNDE3Y2QyYTllNjNmMjBjMGVmZmVkYmYyODg2N2VhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.J2m_Yh7hGgZJgESacZE_ilDgfxG4rS7Luq-syY7GoxI">
> The other issue is that, unless the current change is on the last line, the last line of the file is not shown in the preview. I have a fix for both issues ready to go. Should I push another commit to this branch, or do you want a new PR?

another commit is fine (please include regression tests); I can just re-import since we haven't landed yet.

nk9 committed:
Switch to using iter_inline_changes()

nk9 committed:
Remove orphaned tests.

nk9 commented:
I spent some time on this yesterday. First of all, [here's the code so far](https://github.com/nk9/fastmod/commit/f2b3bcdf0665c2bad0849995494e8ca6d1f5e7d1). In addition to the two things I noticed before, I discovered another serious issue with the existing PR. I was checking specifically for a removal line followed by an insertion line, so the output is wrong if the regex involves mulitpile lines or if the replacement line is removed entirely.

`similar` has code to make this simple, which I've adopted in these new commits. This allowed me to remove a lot of code, as you'll see. However:

- Currently, it seems that the [recommended simple inline diff approach](https://github.com/mitsuhiko/similar/blob/a169e29954502d67e26c9e766f2f719e99a52559/examples/terminal-inline.rs#L30) only lets you show emphasized _words_, not the specific _characters_ which have changed within those words. My original PR highlights characters, and I think that's what I'd prefer if possible. Are we on the same page there? I have [opened an issue](https://github.com/mitsuhiko/similar/issues/84) about this. I'm hoping there's an easy way to do character-level diffs which I haven't noticed yet.
- Instead of all the manual calculation in `diffs_to_print`, I've relied on similar's [`grouped_ops()`](https://docs.rs/similar/2.7.0/similar/struct.TextDiff.html#method.grouped_ops) function. The docs don't say, but the `usize` you pass appears to be [the number of context lines](https://github.com/mitsuhiko/similar/blob/a169e29954502d67e26c9e766f2f719e99a52559/src/common.rs#L103) above or below the change. So just [passing half of the calculated terminal size](https://github.com/nk9/fastmod/commit/f2b3bcdf0665c2bad0849995494e8ca6d1f5e7d1#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR485) seems to behave like the old code. But it isn't exactly the same, so I wanted to check if this is close enough.

You asked about tests, and I had two questions on that.

- Since the `diffs_to_print` function has been removed in favor of the `similar` equivalent, I've had to remove the two tests which rely on `diffs_to_print` (`test_diff_with_unchanged_line_in_middle` and `test_diff_no_changes`). No point testing the library's code in `fastmod`. Let me know your thoughts on this.
- You wanted me to write some tests for the new code, which seems like a good idea. However, I don't see an existing test which actually checks the output printed to the console. All of my changes have been at the point of display, so I don't know how you'd like to test that. The closest the old tests get is to test `diffs_to_print` with various inputs, but that's gone now, and wouldn't have caught any of the issues I've identified anyway. Please note that, quite inconveniently, it seems to not be possible to [pass a `TextDiff` struct around](https://github.com/mitsuhiko/similar/issues/65#issuecomment-3399351055).

I'm sorry this little PR has ballooned, but I'm quite happy you haven't done a release yet. I'm hoping adopting `similar` will end up being a nice simplification, but there are just a few things still to resolve.

nk9 commented:
Well, it looks like `similar` isn't very active, so I don't hold out much hope of getting a response. I've been using my changes for the past few months and it's been working well. Character-level diffs would be nice, but word-level bolding is a big step forward. I don't think users should have to wait for this improvement, I can always come back and revise it if we learn more later.

If you'd like to let me know your thoughts on the request for tests, hopefully we can get this in a build soon!

nk9 commented:
Wanted to pop my head in and give a friendly ping. I'm still loving my local build with the word bolding, really helps draw the eye.

nk9 commented:
Big news: `similar` has [implemented character-level diffs](https://github.com/mitsuhiko/similar/issues/84)! ~~Now it's just a matter of waiting for it to be released.~~ The maintainer has released this in version 3.0.0 today.

I'll have a look at adopting this.

nk9 committed:
Adopt character-level diffs from similar 3.0.0

nk9 committed:
Fix build warnings.

nk9 committed:
Refactor to allow testing the character-level diff output

nk9 commented:
I've adopted the new API, which gives us character-level diffs for free. I've also confirmed that the last-line issues with the original homegrown approach don't affect the similar version. And I've refactored my changes slightly so that it's possible to test the new code, separate from the `similar` code. Unit tests added.

swolchok commented:
FYI I am no longer able to merge PRs to this project.

nk9 commented:
Oh… Alright. Looks like @adamjernst is the new sheriff in town?
